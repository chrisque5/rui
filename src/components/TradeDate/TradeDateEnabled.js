import PropTypes from 'prop-types';
import { InfoCircleFilled } from '@ant-design/icons';
import {
  Checkbox, Form, Tooltip,
} from 'antd';

const TradeDateEnabled = ({
  id, onChange,
}) => (
  <Form.Item
    className="trade-date-enabled"
    valuePropName="checked"
    name={id}
    initialValue={false}
  >
    <Checkbox onChange={onChange} data-testid={`chk-${id}`}>
      <span className="label-text">Override Trade Date</span>
      <Tooltip
        title="Override Trade Date/Time"
        mouseEnterDelay={0.3}
        overlayStyle={{ fontSize: '0.8em', padding: 0, margin: 0 }}
      >
        <InfoCircleFilled />
      </Tooltip>

    </Checkbox>
  </Form.Item>

);

TradeDateEnabled.propTypes = {
  id: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default TradeDateEnabled;
