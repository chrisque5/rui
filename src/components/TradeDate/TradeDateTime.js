import PropTypes from 'prop-types';
import { TimePicker, Form } from 'antd';

const TradeDateTime = ({
  isTradeTimeEnabled, id,
}) => (
  <Form.Item
    className="item-trade-time"
    name={id}
    rules={[{ required: isTradeTimeEnabled, message: 'Trade time required' }]}
  >

    <TimePicker
      disabled={!isTradeTimeEnabled}
      style={{ width: '100%' }}
      placeholder="Trade Time"
      onChange={(time) => time.local()}
    />
  </Form.Item>
);

TradeDateTime.propTypes = {
  isTradeTimeEnabled: PropTypes.bool.isRequired,
  id: PropTypes.string.isRequired,
};

export default TradeDateTime;
