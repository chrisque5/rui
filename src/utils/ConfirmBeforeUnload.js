import { useEffect, useCallback } from 'react';

const ConfirmBeforeUnload = () => {
  const onBeforeUnload = useCallback((e) => {
    e.preventDefault();
    e.returnValue = '';
  }, []);

  useEffect(() => {
    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [onBeforeUnload]);

  return null;
};

export default ConfirmBeforeUnload;
