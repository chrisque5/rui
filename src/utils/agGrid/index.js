import {
  approvalsCellRenderer, approvalsSummaryCellRenderer, isUnderInvestigationCellRenderer, stpCellRenderer, stpFilterCellRenderer,
} from './js-renderers';
import {
  dateFormatter, formatDate, numberFormatter, timeFormatter,
  valueDateFormatter, fixingDateFormatter, publishDateFormatter,
  priceFormatter, brokerageFormatter,
} from './formatters';
import { dateFilterComparator, numberFilterComparitor, timeFilterComparator } from './comparitors';
import { getExportMenuItems } from './exportUtils';

export {
  approvalsCellRenderer,
  approvalsSummaryCellRenderer,
  dateFilterComparator,
  dateFormatter,
  formatDate,
  getExportMenuItems,
  isUnderInvestigationCellRenderer,
  numberFilterComparitor,
  numberFormatter,
  stpCellRenderer,
  stpFilterCellRenderer,
  timeFilterComparator,
  timeFormatter,
  valueDateFormatter,
  fixingDateFormatter,
  publishDateFormatter,
  priceFormatter,
  brokerageFormatter,
};
