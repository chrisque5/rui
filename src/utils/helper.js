import moment from 'moment';
import * as math from 'mathjs';
import { Modal } from 'antd';

export const isValidTerm = (termToTest, selections) => {
  if (termToTest.length < 2) return false;

  const lowerCaseTermToTest = termToTest.toLowerCase();

  const endsWithPeriod = (term) => {
    const periods = ['d', 'w', 'm', 'y'];
    return (periods.includes(term.charAt(term.length - 1)));
  };

  const firstCharacterIsNumber = (term) => {
    const result = term.substring(0, term.length - 1).match(/^[0-9]*$/);
    return result ? result.length > 0 : false;
  };

  const isStaticSelectionValid = (term) => {
    const validStaticTerms = selections.map((x) => x.toLowerCase());
    if (validStaticTerms.includes(term)) {
      return true;
    }

    if (term.startsWith('bmf') || term.startsWith('imm')) {
      let termDaysOnly = term.replace('bmf', '');
      termDaysOnly = termDaysOnly.replace('imm', '');
      const beforeConversion = termDaysOnly;
      termDaysOnly = parseInt(termDaysOnly, 10);
      if (beforeConversion.length === termDaysOnly.toString().length) {
        return (termDaysOnly % 1 === 0 && termDaysOnly > 0 && termDaysOnly < 101);
      }
    }
    return false;
  };

  return isStaticSelectionValid(lowerCaseTermToTest)
    || (endsWithPeriod(lowerCaseTermToTest) && firstCharacterIsNumber(lowerCaseTermToTest));
};

export const doesArrayExistWithValue = (array, index = 0) => {
  if (Array.isArray(array)) {
    if (array && array.length > 0 && array[index]) {
      return true;
    }
  }
  return false;
};

export const setTodayOnNull = (form, id, value) => {
  if (!value) {
    setTimeout(() => {
      form.setFieldsValue({ [id]: moment() });
    });
  }
};

export const dayDifferenceIgnoreHours = (earlierDateMoment, laterDateMoment) => {
  const date1 = earlierDateMoment.clone().startOf('day');
  const date2 = laterDateMoment.clone().startOf('day');
  const dayCount = date1.diff(date2, 'days');

  return dayCount;
};

export const getDateInLocalTZ = () => moment().tz(moment.tz.guess(true)).format('YYYY-MM-DD');
export const getDateTimeInLocalTZ = () => moment().tz(moment.tz.guess(true));

export const parseValueAsBigNumber = (value) => {
  try {
    return math.bignumber(value);
  } catch (err) {
    return null;
  }
};

export const formatAmount = (val) => {
  const inputValue = val.replace(/\s/g, '');
  let multiplier = 1;
  const oneM = 1000000;
  const oneB = 1000000000;
  let amt = inputValue;

  if (Number.isNaN(Number(inputValue)) && inputValue.length > 1) {
    const lastChar = inputValue.charAt(inputValue.length - 1).toLowerCase();
    amt = Number(inputValue.substring(0, inputValue.length - 1));
    if (lastChar === 'm') {
      multiplier = oneM;
    } else if (lastChar === 'b') {
      multiplier = oneB;
    } else {
      return inputValue;
    }
  } else if (Number(inputValue) < oneM) {
    multiplier = oneM;
  }

  const calculatedAmount = math.multiply(math.bignumber(amt), math.bignumber(multiplier));
  return calculatedAmount.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const isValidDate = (date, dateFormat) => moment(date, dateFormat || 'L', true).isValid();

export const findCurrencyPair = (ccy1, ccy2, currencies) => (
  currencies.find(
    (currency) => currency.baseCurrency === ccy1 && currency.counterCurrency === ccy2,
  )
);

export const filterSpotExecutionVenues = (execVenues) => execVenues.filter((i) => i.venueId === 'XOFF');

export const isDealTypeSame = (dealType1, dealType2) => (dealType1 === dealType2);

export const isStrategyTypeSame = (strategyType1, strategyType2) => (strategyType1 === strategyType2);

export const ConfirmUnsavedChanges = (message, onYes, onNo) => {
  Modal.confirm({
    title: 'Confirm',
    content: message,
    okText: 'Yes',
    cancelText: 'No',
    onOk: () => {
      if (onYes) {
        setTimeout(() => {
          onYes();
        });
      }
    },
    onCancel: () => {
      if (onNo) {
        setTimeout(() => {
          onNo();
        });
      }
    },
  });
};

export const isObjectEmpty = (obj) => Object.keys(obj).length === 0 && obj.constructor === Object;

export const getUniqueID = () => {
  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
  return `${s4() + s4()}-${s4()}`;
};

export const showPreviousValueToolTip = (currentValue, previousValue) => {
  if (currentValue !== previousValue) {
    return { title: `Previous Value: ${previousValue || 'N/A'}`, className: 'deal-input-dirty' };
  }
  return { title: '', className: '' };
};
