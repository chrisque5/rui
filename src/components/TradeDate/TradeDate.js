import PropTypes from 'prop-types';
import { DatePicker, Form } from 'antd';
import moment from 'moment';
import { setTodayOnNull } from '../../utils/helper';

const TradeDate = ({
  form, tradeDateOnChange, isTradeDateEnabled, id,
}) => (
  <Form.Item
    className="item-trade-date"
    name={id}
    rules={[{ required: isTradeDateEnabled, message: 'Trade date required' }]}
  >

    <DatePicker
      suffixIcon
      disabled={!isTradeDateEnabled}
      format="L"
      onChange={(value) => { setTodayOnNull(form, id, value); tradeDateOnChange(value, id); }}
      allowClear={false}
      disabledDate={(current) => current > moment().endOf('day')}
      style={{ paddingRight: '5px', width: '95%' }}
      placeholder="Trade Date"
    />
  </Form.Item>
);

TradeDate.propTypes = {
  form: PropTypes.shape().isRequired,
  isTradeDateEnabled: PropTypes.bool.isRequired,
  tradeDateOnChange: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
};

export default TradeDate;
