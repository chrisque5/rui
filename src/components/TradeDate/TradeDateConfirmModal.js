import { ExclamationCircleFilled } from '@ant-design/icons';
import { Button, Modal } from 'antd';
import { useDispatch } from 'react-redux';
import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { ids } from '../../utils/constants';
import { toggleTradeDateSubmitPopup } from '../../actions/uiActions';
import { getDateTimeInLocalTZ } from '../../utils/helper';

const TradeDateConfirmModal = ({
  form, isVisible, onAccept,
}) => {
  const dispatch = useDispatch();
  const [tradeDate, setTradeDate] = useState();
  const [tradeTime, setTradeTime] = useState();
  const [currentDate, setCurrentDate] = useState(getDateTimeInLocalTZ());

  useEffect(() => {
    if (isVisible) {
      setCurrentDate(getDateTimeInLocalTZ());
    }
  }, [isVisible]);

  useEffect(() => {
    const formTradeDate = form.getFieldValue(ids.TRADE_DATE);
    const formTradeTime = form.getFieldValue(ids.TRADE_TIME);

    if (formTradeDate) {
      setTradeDate(moment(formTradeDate).format('Do MMMM YYYY'));
    }

    if (formTradeTime) {
      setTradeTime(moment(formTradeTime).format('HH:mm:ss'));
    }
  }, [form, isVisible]);

  const closePopup = () => {
    dispatch(toggleTradeDateSubmitPopup(false));
  };

  const onclick = () => {
    closePopup();
    onAccept();
  };

  return (
    <Modal
      className="confirm-modal"
      title="Confirm Trade Date/Time"
      visible={isVisible}
      onCancel={closePopup}
      bodyStyle={{ padding: '10px' }}
      footer={[
        <Button data-testid="trade-date-modal-cancel" key="back" onClick={closePopup}>Cancel</Button>,
        <Button data-testid="trade-date-modal-save" key="submit" type="primary" onClick={onclick}>Accept</Button>,
      ]}
      destroyOnClose
    >
      <div className="modal-body">
        <span className="modal-header">
          <ExclamationCircleFilled />
          Trade Date/Time has been amended.
        </span>
        <span className="modal-times">
          <span>Current Trade Date/Time: </span>
          <span>{`${moment(currentDate).format('Do MMMM YYYY')} ${moment(currentDate).format('HH:mm:ss')}`}</span>
        </span>
        <span className="modal-times">
          <span>Amended Trade Date/Time:  </span>
          <span>{`${tradeDate} ${tradeTime}`}</span>
        </span>
      </div>
    </Modal>
  );
};

TradeDateConfirmModal.propTypes = {
  form: PropTypes.shape().isRequired,
  isVisible: PropTypes.bool.isRequired,
  onAccept: PropTypes.func.isRequired,
};

export default TradeDateConfirmModal;
