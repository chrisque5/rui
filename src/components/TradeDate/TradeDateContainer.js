import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { ids } from '../../utils/constants';
import {
  getIsTradeDateSubmitModalVisible, getIsTradeDateEnabled, getUserPreferenceSettingsRatesFeed,
} from '../../utils/selectors';
import { toggleIsTradeDateEnabled } from '../../actions/uiActions';
import TradeDate from './TradeDate';
import TradeDateTime from './TradeDateTime';
import TradeDateEnabled from './TradeDateEnabled';
import TradeDateConfirmModal from './TradeDateConfirmModal';
import { getDateTimeInLocalTZ } from '../../utils/helper';
import { showWarningNotification } from '../../utils/notifications';

const TradeDateContainer = ({
  form, tradeDateOnChange, submitDealWithCustomDate,
}) => {
  const dispatch = useDispatch();
  const [initialTime, setInitialTime] = useState(null);
  const isTradeDateConfirmModalVisible = useSelector(getIsTradeDateSubmitModalVisible);
  const isTradeDateEnabled = useSelector(getIsTradeDateEnabled);
  const isRatesFeedEnabled = useSelector(getUserPreferenceSettingsRatesFeed);

  const onCheckBoxChange = (e) => {
    const checked = e.target.checked;
    dispatch(toggleIsTradeDateEnabled(checked));
    // if checkbox toggled on initialise the time
    if (checked) {
      const currentTime = getDateTimeInLocalTZ();
      setInitialTime(currentTime);
      if (form) {
        form.setFieldsValue({
          [ids.TRADE_DATE]: currentTime,
          [ids.TRADE_TIME]: currentTime,
        });
      }
      if (isRatesFeedEnabled) {
        showWarningNotification(
          'Rate Error',
          'Rate feed is not available when overriding trade date/time. Please set rates manually.',
        );
      }
    } else {
      // if checkbox toggled off reset the time
      setInitialTime(null);
      if (form) {
        form.resetFields([ids.TRADE_DATE, ids.TRADE_TIME]);
      }
      tradeDateOnChange();
    }
  };

  return (
    <>
      <TradeDateEnabled form={form} id={ids.IS_TRADE_DATE_ENABLED} onChange={onCheckBoxChange} />
      <TradeDate
        form={form}
        isTradeDateEnabled={isTradeDateEnabled}
        initialTime={initialTime}
        id={ids.TRADE_DATE}
        tradeDateOnChange={tradeDateOnChange}
      />
      <TradeDateTime
        form={form}
        isTradeTimeEnabled={isTradeDateEnabled}
        initialTime={initialTime}
        id={ids.TRADE_TIME}
      />
      <TradeDateConfirmModal form={form} onAccept={submitDealWithCustomDate} isVisible={isTradeDateConfirmModalVisible} />
    </>
  );
};

TradeDateContainer.propTypes = {
  form: PropTypes.shape().isRequired,
  submitDealWithCustomDate: PropTypes.func.isRequired,
  tradeDateOnChange: PropTypes.func.isRequired,
};

export default TradeDateContainer;
