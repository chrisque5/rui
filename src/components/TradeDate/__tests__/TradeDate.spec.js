import {
  renderWithForm, waitFor, fireEvent, getByTitle,
} from 'test-utils/testUtils';
import * as helpers from '../../../utils/helper';
import TradeDate from '../TradeDate';

jest.mock('moment', () => () => jest.requireActual('moment')('2020-07-22T00:00:00.000Z'));

const mockForm = {};
const tradeDateOnChange = jest.fn();
const defaultProps = {
  form: mockForm,
  id: 'tradeDate',
  isTradeDateEnabled: false,
  initialTime: '01012020',
  tradeDateOnChange,
};
describe('<TradeDate/>', () => {
  const renderComponent = () => renderWithForm(<TradeDate {...defaultProps} />);
  const setTodayOnNullSpy = jest.spyOn(helpers, 'setTodayOnNull');
  test('<TradeDate /> renders snapshot as expected', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('<TradeDate /> with prop.isTradeDateEnabled = false renders the DatePicker disabled', async () => {
    const { getByPlaceholderText } = renderComponent();
    const datePicker = await waitFor(() => getByPlaceholderText('Trade Date'));
    expect(datePicker.disabled).toBe(true);
  });

  test('<TradeDate /> with prop.isTradeDateEnabled = true renders the DatePicker enabled', async () => {
    const props = { ...defaultProps, isTradeDateEnabled: true };
    const { getByPlaceholderText } = renderWithForm(<TradeDate {...props} />);
    const datePicker = await waitFor(() => getByPlaceholderText('Trade Date'));
    expect(datePicker.disabled).toBe(false);
  });

  test('<TradeDate /> calls onChange functions when date selected', async () => {
    const props = { ...defaultProps, isTradeDateEnabled: true };
    const { getByPlaceholderText } = renderWithForm(<TradeDate {...props} />);

    const datePicker = await waitFor(() => getByPlaceholderText('Trade Date'));
    fireEvent.mouseDown(datePicker);

    const selectedDateElem = await waitFor(() => getByTitle(document.body, '2020-07-22'));
    fireEvent.click(selectedDateElem);

    expect(tradeDateOnChange).toHaveBeenCalledTimes(1);
    expect(setTodayOnNullSpy).toHaveBeenCalledTimes(1);
  });
});
