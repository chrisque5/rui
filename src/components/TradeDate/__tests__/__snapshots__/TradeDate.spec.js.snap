// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TradeDate/> <TradeDate /> renders snapshot as expected 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item item-trade-date"
    >
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-picker ant-picker-disabled"
              style="padding-right: 5px; width: 95%;"
            >
              <div
                class="ant-picker-input"
              >
                <input
                  autocomplete="off"
                  disabled=""
                  id="tradeDate"
                  placeholder="Trade Date"
                  readonly=""
                  size="12"
                  title=""
                  value=""
                />
                <span
                  class="ant-picker-suffix"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
